[chore][fileconsumer/archive] - add signature methods for archive feature

VihasMakwana opened the pull request:
Description:

This PR introduces the `archive` method to store files older than 3 poll cycles.
The core logic for reading will be introduced in a follow up PR.


Link to tracking Issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32727

Testing: To be included in future PRs

Documentation: To be included in future PRs

VihasMakwana committed:
chore: add initial structure and signature methods

VihasMakwana committed:
fix: add license

VihasMakwana committed:
fix: lint

VihasMakwana committed:
fix: lint

VihasMakwana committed:
fix: check

VihasMakwana committed:
chore: add write method, update interface

VihasMakwana committed:
Merge branch 'main' into create-archive-storage

VihasMakwana committed:
chore: commit, archive module

VihasMakwana commented:
@djaglowski This is a draft PR which separates the reader creation and archiving module as per https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32727#issuecomment-2339665263

VihasMakwana commented on pkg/stanza/fileconsumer/internal/archive/archive.go:
```diff
@@ -0,0 +1,73 @@
+
+type FileRecord struct {
+	File        *os.File
+	Fingerprint *fingerprint.Fingerprint
```
@djaglowski storing file handle is crucial for us because of the way we handle the archiving:

1. We try to match files against `knownFiles` array.
2. If a match isn't found, we would accumulate the unmatched files (instead of finding in the archive. More about it was discussed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32727#issuecomment-2260891596).
3. Perform the archiving logic. Described [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/431ab4163432fe03658ddd0f34e89598db4abad7/pkg/stanza/fileconsumer/internal/archive/archive.go#L51-L59).

We need the file handle so that we can create a reader from it after the archiving is done.

djaglowski replied:
This doesn't make any sense to me. See my comments on the other PR. I think this is a fundamental point that needs to be resolved before bothering with the rest of the PR.

`reader.Metadata` is sufficient and necessary for creating readers in all other circumstances _except where we already have an open file handle_, in which case we can do better. 

I do not understand why we would or even could have open file handles at this point, since the archive should only be populated by elements which are discarded from `knownFiles`, which only contains `reader.Metadata`.

VihasMakwana committed:
chore: cleanup, simplify the PR

VihasMakwana commented:
@djaglowski please take a look at this PR. Appreciate your time reviewing this!
 
--- 
Addressing your points below:

> What does storing the file handle give us?

As I said, I made things too complex with the initial PR, and we don’t actually need to keep files open while archiving. 

> Are we sure archive should be separate from tracker? Seems very closely related and kind of awkward to manage both.

Agree. Keeping everything in tracker would make it more maintainable. 

> reader.Metadata is sufficient and necessary for creating readers in all other circumstances except where we already have an open file handle, in which case we can do better.

This makes sense to me.

VihasMakwana committed:
fix: initial commit, second PR

VihasMakwana committed:
chore: remove stanza.go

VihasMakwana committed:
Format the comment

VihasMakwana committed:
chore: use options

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -110,9 +116,16 @@ func (t *fileTracker) ClosePreviousFiles() (filesClosed int) {
 	return
 }
 
+func (t *fileTracker) SetPersister(persister operator.Persister) {
```
Let's pass this in during instantiation so that it is not an effect that can be applied later.

djaglowski replied:
Also because it may not be a concern of other implementation of tracker.

VihasMakwana replied:
@djaglowski we instantiate the tracker while building the config. The `persister` is not available to us in that part of code. It's available in `Start(...)`.

djaglowski reviewed (commented):
@VihasMakwana, sorry for the delay. This looks good to me but I'd like to see more before merging complexity into the package. Can we expand this PR to include the implementation?

VihasMakwana commented:
@djaglowski yes, we can do that. I raised a [DRAFT PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35216) which includes the implementation. I'll merge that in this PR.

VihasMakwana commented:
@djaglowski this PR includes the implementation now! 

I'll raise a follow up PR which covers the **_reading from archive_** part

VihasMakwana committed:
fix: lint

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana committed:
fix: bug

VihasMakwana committed:
chore: rename function

VihasMakwana committed:
chore: rename function

VihasMakwana commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -26,6 +31,7 @@ type Tracker interface {
 	EndPoll()
 	EndConsume() int
 	TotalReaders() int
+	EnableArchiving(persister operator.Persister)
```
@djaglowski I've renamed this function. It sounds much better to me. Let me know how it looks to you.

djaglowski replied:
It's not the name that worries me really. The problem is that instantiating the tracker and then having to reconfigure it later. Maybe the tracker should be instantiated entirely in Start?

VihasMakwana replied:
Makes sense. Opened a new PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35098 to tackle this, as they're independent changes.

VihasMakwana replied:
@djaglowski I meant https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35492. Sorry 🥲

djaglowski replied:
I don't see them as unrelated changes. The reason I suggested instantiating the tracker in `Start` is so that the persister can be passed to it when it is constructed. I don't see any point to #35492 without the archive.

VihasMakwana replied:
Ah, ok. Let me merge both in this then and we can get it done. Thanks!

VihasMakwana commented:
@djaglowski the CI looks good now.

Also, I've left testing for now and I'll add it in the follow up PR>

VihasMakwana commented:
/label changelog

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

djaglowski commented on pkg/stanza/fileconsumer/file.go:
```diff
@@ -58,6 +59,11 @@ func (m *Manager) Start(persister operator.Persister) error {
+			m.tracker.EnableArchiving(persister)
+		}
+	} else if m.pollsToArchive > 0 {
+		return fmt.Errorf("archiving is not supported in memory, please use a storage extension")
```
Returning an error during Start will crash the entire collector. I think we either need to just log an error, or add this to the polls we remember in memory. For now, I would just log an error and we can worry about more complicated solutions later if needed.

VihasMakwana replied:
That works. Thanks!

VihasMakwana replied:
Done.

VihasMakwana committed:
chore: cleanup tracker and use options

VihasMakwana committed:
fix: move function before loading

VihasMakwana committed:
chore: lint, ci

VihasMakwana committed:
chore: log the error

VihasMakwana committed:
chore: remove redundant argument

VihasMakwana committed:
Merge branch 'cleanup-tracker-options' into create-archive-storage-temp

VihasMakwana committed:
chore: remove redundant code

VihasMakwana commented:
@djaglowski I've moved the instantiation of tracker in this PR. PTAL!

VihasMakwana committed:
chore: add new no tracking

djaglowski commented on pkg/stanza/fileconsumer/file.go:
```diff
@@ -47,6 +49,9 @@ func (m *Manager) Start(persister operator.Persister) error {
 	}
 
+	// instantiate the tracker
+	m.instantiateTracker()
```
```suggestion
	m.instantiateTracker(persister)
```

VihasMakwana replied:
Missed that. Thanks!

VihasMakwana committed:
fix: pass persister instead of m.persister

djaglowski commented on pkg/stanza/fileconsumer/file.go:
```diff
@@ -261,3 +270,14 @@ func (m *Manager) newReader(ctx context.Context, file *os.File, fp *fingerprint.
 }
+
+func (m *Manager) instantiateTracker(persister operator.Persister) {
+	opts := []tracker.OptionFunc{tracker.WithMaxBatchFiles(m.maxBatchFiles)}
+	if m.noTracking {
+		opts = append(opts, tracker.WithNoTracking())
+	}
+	if m.pollsToArchive > 0 {
+		opts = append(opts, tracker.WithPollsToArchive(m.pollsToArchive), tracker.WithPersister(persister))
+	}
+	m.tracker = tracker.NewFileTracker(m.set, opts...)
```
It seems unnecessary to change the pattern we already had. If there's no tracking, just instantiate a `NewNoStateTracker`.

If there is tracking, just pass it the batch size and persister (even if it is nil). I don't think there's enough complexity to justify the tracker having its own options yet.

Inside the tracker, we can make decisions about archiving based on whether the persister is nil.

VihasMakwana replied:
Yeah, that works. It's probably the best option considering and we can change the pattern in future, if needed.

VihasMakwana replied:
Done. Looks much simple. Thanks!

VihasMakwana committed:
chore: remove options

VihasMakwana committed:
chore: fix tests

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana commented:
@djaglowski Thanks for your time in reviewing this PR!
Once this gets merged, I'll roll out another PR with **_reading from archive_** logic.
It'll also contain the test cases for this feature.

We can ofcourse break it down in multiple PRs if necessary.

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -125,6 +143,34 @@ func (t *fileTracker) TotalReaders() int {
+	//                   start
+	//                   index
+
+	if t.pollsToArchive <= 0 || t.persister == nil {
+		return
+	}
```
Do we have to worry about these cases? Aren't we checking them and instantiating a `NoStateTracker` when these would ever be relevant?

VihasMakwana replied:
`NoStateTracker` will be created when `WithNoTracking` option is selected. 
We can create a normal tracker without archiving disabled, by setting `pollsToArchive: 0`.
A tracker with archiving enabled will have `pollsToArchive` > `0`. 

The check `t.persister == nil` is just a precautionary check.

djaglowski replied:
Alright. Maybe in another PR we can clean this up. IMO we have starting to have too many special cases to handle and the easiest solution to that is move to a pattern where the nils pieces are nops instead. Then the implementation doesn't need to worry about details so much.

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -125,6 +143,34 @@ func (t *fileTracker) TotalReaders() int {
+		return
+	}
+	key := fmt.Sprintf("knownFiles%d", t.archiveIndex)
+	if err := checkpoint.SaveKey(context.Background(), t.persister, metadata.Get(), key); err != nil {
```
Can we just checkpoint in exactly the same place as we were before? I'm confused why we need to add a separate milestone for this.

VihasMakwana replied:
@djaglowski I’ve made some adjustments and relocated the archive checkpointing to just below where we perform our standard checkpointing.

djaglowski replied:
I did, but reconsidering this comment - when using an archive, we have to load each set from storage separately, so I see we also have to write it that way.

VihasMakwana committed:
chore: move checkpointing to file.go

VihasMakwana commented on pkg/stanza/fileconsumer/file.go:
```diff
@@ -73,7 +80,9 @@ func (m *Manager) Stop() error {
 	}
 	m.wg.Wait()
-	m.telemetryBuilder.FileconsumerOpenFiles.Add(context.TODO(), int64(0-m.tracker.ClosePreviousFiles()))
+	if m.tracker != nil {
```
As it turns out, we don't need to pass persister into tracker. 

I've introduced a new method `GetArchiveMetadata`, which returns the metadata to be stored in archive (i.e. `knownFiles[2]`). It looks much simpler now. PTAL!

VihasMakwana replied:
Since we initialize the tracker in `Start`, the component lifecycle tests for `otlpjsonfilereceiver` are causing panics due to `Shutdown` being called first. We need to implement a check for this.

VihasMakwana replied:
Same reason as we check for `m.cancel` here
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/985dddba71e26212e7e0535965fbec205135958f/pkg/stanza/fileconsumer/file.go#L71-L74

VihasMakwana committed:
Revert "chore: move checkpointing to file.go"

This reverts commit 27cc4e1b017a2d9ee0af499cbea823b5dd4d030a.
